src/ssl.c

```diff
@@ -51319,6 +51342,608 @@ int wolfSSL_PEM_write_bio_PKCS7(WOLFSSL_BIO* bio, PKCS7* p7)
+            section[sectionLen+1] = '\0';
+
+            *bcont = wolfSSL_BIO_new(wolfSSL_BIO_s_mem());
+            ret = wolfSSL_BIO_write(*bcont, section, (int)XSTRLEN(section));
+            if (ret != (int)XSTRLEN(section)) {
+                MIME_free_hdrs(allHdrs);
+                XFREE(boundary, NULL, DYNAMIC_TYPE_PKCS7);
+                XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+                wolfSSL_BIO_free(*bcont);
+                return NULL;
+            }
+            if ((bcontMemSz = wolfSSL_BIO_get_mem_data(*bcont, &bcontMem)) < 0) {
+                MIME_free_hdrs(allHdrs);
+                XFREE(boundary, NULL, DYNAMIC_TYPE_PKCS7);
+                XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+                wolfSSL_BIO_free(*bcont);
+                return NULL;
+            }
+
+
+            MIME_free_hdrs(allHdrs);
+            ret = MIME_parse_headers(in, remainLen, &allHdrs);
+            if (ret < 0) {
+                WOLFSSL_MSG("Parsing MIME headers failed.\n");
+                MIME_free_hdrs(allHdrs);
+                XFREE(boundary, NULL, DYNAMIC_TYPE_PKCS7);
+                XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+                wolfSSL_BIO_free(*bcont);
+                return NULL;
+            }
+            curHdr = MIME_find_header_name("Content-Type", allHdrs);
+            if (!curHdr || (XSTRNCMP(curHdr->body,
+                            "application/pkcs7-signature",
+                            XSTR_SIZEOF("application/pkcs7-signature")) &&
+                            XSTRNCMP(curHdr->body,
+                            "application/x-pkcs7-signature",
+                            XSTR_SIZEOF("application/x-pkcs7-signature")))) {
+                WOLFSSL_MSG("S/MIME headers not found inside "
+                            "multipart message.\n");
+                MIME_free_hdrs(allHdrs);
+                XFREE(boundary, NULL, DYNAMIC_TYPE_PKCS7);
+                XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+                wolfSSL_BIO_free(*bcont);
+                return NULL;
+            }
+
+            section[0] = '\0';
+            sectionLen = 0;
+            lineLen = wolfSSL_BIO_gets(in, section, remainLen);
+            while(XSTRNCMP(&section[sectionLen], boundary, boundLen) &&
+                  remainLen > 0) {
+                sectionLen += lineLen;
+                remainLen -= lineLen;
+                lineLen = wolfSSL_BIO_gets(in, &section[sectionLen],
+                                           remainLen);
+                if (lineLen<=0) {
+                    MIME_free_hdrs(allHdrs);
+                    XFREE(boundary, NULL, DYNAMIC_TYPE_PKCS7);
+                    XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+                    wolfSSL_BIO_free(*bcont);
+                    return NULL;
+                }
+            }
+
+            XFREE(boundary, NULL, DYNAMIC_TYPE_PKCS7);
+        }
+    }
+    else if (curHdr && (!XSTRNCMP(curHdr->body, "application/pkcs7-mime",
```

Comment:
Consider having all string as `static const char` at top of function or file. Example: `static const char* kPkcs7AppMime = "application/pkcs7-mime";` and referencing it instead of duplicating the string. Same for the other string instances used in this function.